utilities/merge_operators/int64subtract.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include <memory>
+
+#include "logging/logging.h"
+#include "rocksdb/env.h"
+#include "rocksdb/merge_operator.h"
+#include "rocksdb/slice.h"
+#include "util/coding.h"
+#include "utilities/merge_operators.h"
+
+using namespace ROCKSDB_NAMESPACE;
+
+namespace {  // anonymous namespace
+
+// A 'model' merge operator with int64 subtraction semantics
+// operands and database value should be variable length encoded
+// int64_t values, as encoded/decoded by `util/coding.h`.
+class Int64SubtractMergeOperator : public AssociativeMergeOperator {
```

Comment:
It shouldn't inherit from `AssociativeMergeOperator` because subtraction is non-associative, e.g., `5 - (4 - 3) != (5 - 4) - 3`.

Reply:
Or closer to the implementation:

    ((/*implicit*/ 0 - 5) - 4) - 3

!=

    (/*implicit*/ 0 - 5) - ((/*implicit*/ 0 - 4) - 3)

For an operator to be associative, you have to be able to "bootstrap" from any input element without messing up the merge semantics. I think it's completely misleading, if not wrong, that the operands to AssociativeMergeOperator::Merge are called `existing_value` and `value` rather than `left` (optional) and `right`.

Reply:
@pdillinger Good point. Should I send a separate PR to rename the existing ones to `left` and `right`?

Reply:
@adamretter I'll take it to put in my preferred wording ;)